[FEAT] 카카오 소셜 로그인을 위한 서버 로직 작성

jye326 opened the pull request:
## Issue Number
closed #801 

## As-Is
<!-- 문제 상황 정의 -->

소셜 로그인 API를 추가합니다.

## Check List
- [x] 테스트가 전부 통과되었나요?
- [x] 모든 commit이 push 되었나요?
- [x] merge할 branch를 확인했나요?
- [x] Assignee를 지정했나요?
- [x] Label을 지정했나요?
- [x] 닫을 이슈 번호를 지정했나요?


# 카카오 로그인 과정

1. 클라이언트에서 카카오 로그인이 완료되면 서버의 RedirectUrl로 `code`를 포함해 요청
    1. `/kakao/callback?code=codefromkakao` 
2. 서버는 code를 받아서 `https://kauth.kakao.com/oauth/token` 으로 POST
    1. https://developers.kakao.com/docs/latest/ko/kakaologin/rest-api#request-token
    2. 요청 본문
    
    ```sql
    grant_type=authorization_code(고정)
    client_id=서버환경변수.REST_API_KEY
    redirect_uri=서버환경변수.REDIRECT_URI
    code=클라이언트에서 전달받은 code
    client_secret=서버환경변수.CLIENT_SECRET
    // state = 추후 고려, 이슈 참고
    ```
    
    c. `kakaoAccessToken`을 받아온다.
    
3. 서버는 `kakaoAccessToken`으로 https://kapi.kakao.com/v2/user/me 에서 `Long: kakaoId`(카카오의 회원 식별번호)를 받아온다.

### Member_Oauth 테이블을 확인한다.

```sql
CREATE TABLE member_oauth (
  id BIGINT AUTO_INCREMENT PRIMARY KEY,
  member_id BIGINT NOT NULL,
  provider VARCHAR(20) NOT NULL,
  provider_user_id VARCHAR(128) NOT NULL,
  UNIQUE(provider, provider_user_id),
  FOREIGN KEY (member_id) REFERENCES member(id)
);
```

### 1) provider = `'KAKAO'`, provider_user_id = `kakaoId` 인 칼럼이 존재할 경우

1. member_id로 연결된 멤버를 찾아 기존 로그인 로직 대로 로그인한다. (accessToken, refreshToken 발급)

### 2) provider = `'KAKAO'`, provider_user_id = `kakaoId` 인 칼럼이 존재하지 않는 경우

1. **Oauth 회원 가입 페이지**로 리다이랙트 응답을 보낸다.
    1. Oauth 회원가입 유저 식별을 위해 `kakaoId`를 subject로 갖는 `oauthSignUpToken`을 응답 헤더에 추가한다.
2. **Oauth 회원 가입** `/oauth-signup`
    1. **Member 회원가입** : 이름, 성별, 전화번호(인증 포함) 정보와 함께,  `NOT NULL`  인 `loginId`, `password`는 난수로 임의 생성한다.
    2. **MemberOauth 회원가입** : `oauthSignUpToken` 으로 해당 멤버를 Oauth 회원 테이블에 연결한다.
3. **1-1 절차대로 로그인 처리**

---

## 해당 방식을 채택한 근거

1. 기존 Member 테이블 구조와 로그인 및 토큰 확인 구조를 변경하지 않고, 새로운 `Member_Oauth` 테이블만 추가하여 Oauth를 지원할 수 있음
2. KAKAO 이외의 다른 Oauth 제공 플랫폼에 대한 확장성이 높음

## 이슈

- 현재 로그인 과정은 state를 통한 브라우저 인증을 진행하지 않고 있기 때문에 공격자의 CSRF 공격에 취약한 상태
    - 로그인을 시작하기 전에 난수 state 값을 생성하여 클라이언트와 서버가 각각 나눠 가지고, 로그인 과정에서의 state 값 비교를 통해 브라우저의 출처를 확인해야 함.
    - 이때 서버가 state 값을 생성하고 저장하는 경우에는
        - 분산 환경을 고려해 Redis 등의 공유 저장소를 사용하거나
        - DB에 state를 별도로 관리하는 테이블을 생성하거나
        - base64(HMAC+난수+타임스탬프(만료시간)) 값으로 stateless 하게 관리해야 함.
    - 1, 2번이 다소 오바스러워 3번으로 가야 할 것 같은데 시간 관계 상 일단 미뤄두었다.

- redirect_uri를 토큰 교환 단계에서 다시 보내는 건 **OAUTH** **표준 규격 + 보안 검증(코드 탈취 방지)** 때문.
    - “내 서버에 이미 요청이 왔으니까 굳이 필요 없지 않나?” 싶을 수 있지만, **code를 중간에 탈취한 공격자가 다른 redirect_uri로 쓰지 못하게 막기 위해서** 반드시 다시 보내야 함.

jye326 committed:
feat: 카카오 로그인 콜백 엔드포인트 추가

- 카카오 로그인 콜백 처리 위한 `/kakao/callback` 엔드포인트 구현
- 요청 파라미터(code, error 등) 처리 로직 추가
- 에러 및 성공 시 응답 상태 코드 및 메시지 구성

jye326 committed:
feat: 카카오 로그인 로직에 OauthClientService 추가

- AuthService에 OauthClientService 의존성 주입
- Kakao 로그인 처리 메서드 `kakaoLogin` 추가

jye326 committed:
feat: 카카오 로그인 OauthClientService 및 환경 변수 설정 추가

- OauthClientService 클래스 생성 및 Kakao 로그인 관련 메서드 추가
- `KakaoTokenResponse`, `KakaoUserInfoResponse` DTO 추가
- `kakaoLogin` 메서드에 AccessToken 및 사용자 정보 조회 로직 추가
- application.yml 및 CI/CD workflows에 Kakao 관련 환경 변수 설정 추가

jye326 committed:
feat: OAuth 회원가입용 토큰 생성 메서드 추가

- `createOauthSignUpToken` 메서드 구현
- OAuth 회원가입 시 필요한 유효기간 토큰 생성 로직 추가
- `JwtProvider` 클래스에 새로운 메서드 추가 적용

jye326 committed:
feat: OAuth 회원 관리 기능 초기 구현

- `member_oauth` 테이블 생성 SQL 추가
- OAuth 회원 관리용 `MemberOauth` 엔티티 생성
- `MemberOauth` 관련 repository 인터페이스 `MemberOauthRepository` 추가
- OAuth 회원 정보 저장 및 조회 기능 구현 준비

jye326 committed:
feat: 카카오 OAuth 로그인 및 회원가입 기능 추가

- `kakaoLogin` 메서드 구현: OAuth 회원 로그인 처리 및 신규 회원가입 토큰 반환
- OAuth 회원가입 처리 메서드 `registerOauthMember` 추가
- `AuthTokenResponse` 클래스에 OAuth 회원가입 토큰 추가 필드 및 관련 메서드 구현
- 컨트롤러에 `/oauth-signup` API 추가 및 로그인 시 OAuth 토큰 처리 로직 반영
- `OauthSignUpRequest` DTO 생성
- 테스트 코드에 OAuth 클라이언트 및 회원가입 로직 추가
- `build.gradle` 종속성 업데이트: `commons-lang3` 추가

jye326 committed:
fix: OAuth 회원가입 시 기존 쿠키 삭제 로직 추가

- `CookieWriter.clearCookies` 메서드를 통해 이전 쿠키 제거 처리
- OAuth 회원가입 처리 시 새로운 토큰 작성 전 쿠키 초기화 로직 반영
- 기존 쿠키로 인한 토큰 충돌 문제 해결

jye326 committed:
feat: member_oauth 테이블 소프트 삭제 컬럼 추가

- `is_deleted` 컬럼 추가 및 기본값 `FALSE` 설정
- `deleted_at` 컬럼 추가 및 NULL 허용

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/infra/OauthClientService.java:
```diff
@@ -0,0 +1,47 @@
+    private String kakaoRedirectUri;
+    @Value("${kakao.client-id}")
+    private String kakaoClientId;
+    private static final String kakaoUserInfoRequestUri = "https://kapi.kakao.com/v2/user/me";
```
p3: `static` 필드를 위로 올리는 건 어떨까요?

jye326 replied:
반영하였습니다!

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/infra/OauthClientService.java:
```diff
@@ -0,0 +1,47 @@
+                .body(KakaoTokenResponse.class);
+    }
+
+    public KakaoUserInfoResponse requestKakaoId(String accessToken) {
```
p3: 카카오톡에서 발급한 액세스 토큰이라는 것을 알 수 있도록 파라미터 명을 `kakaoAccessToken`으로 하는 건 어떨까요? 그냥 `accessToken`으로 하면 저희 서버에서 발급한 액세스 토큰도 포함될 것 같은 네이밍이라서요!

jye326 replied:
동의합니다! 수정했습니다!

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/MemberOauth.java:
```diff
@@ -0,0 +1,53 @@
+    private Member member;
+
+    @Column(nullable = false)
+    private String provider;
```
p3: `provider`은 Enum으로 만들면 좋을 것 같군요.

ppparkta replied:
p3: 카카오에 한정된 테이블이 아니라 oauth 전체를 커버할 수 있도록 테이블을 짜주신건가요? 좋네요! 프리 말대로 enum으로 관리하면 좋을 것 같아요

jye326 replied:
반영했습니다!

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/AuthService.java:
```diff
@@ -96,4 +111,46 @@ private Member createMember(SignUpRequest request) {
+    public MemberOauth registerOauthMember(@Valid OauthSignUpRequest request) {
+        String oauthSignUpToken = request.oauthSignUpToken();
+        String kakaoId = String.valueOf(jwtProvider.getSubjectFromPayloadBy(oauthSignUpToken));
+        String randomId = RandomStringUtils.randomAlphanumeric(20);
```
p3: `Member`에서 `loginId`라는 명칭을 사용하고 있으니 여기서도 `randomLoginId`로 통일해주면 좋을 것 같습니다!

jye326 replied:
반영했습니다~

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/AuthService.java:
```diff
@@ -96,4 +111,46 @@ private Member createMember(SignUpRequest request) {
+
+    public MemberOauth registerOauthMember(@Valid OauthSignUpRequest request) {
+        String oauthSignUpToken = request.oauthSignUpToken();
+        String kakaoId = String.valueOf(jwtProvider.getSubjectFromPayloadBy(oauthSignUpToken));
```
p3: `kakaoId` 대신 `OauthId` 등을 사용해도 될 것 같습니다. 카카오 로그인에 한정된 게 아니라 OAuth 회원가입 모두 이 로직을 사용할 것 같아서요!

ppparkta replied:
프리 의견 좋네요!

jye326 replied:
반영하였습니다! 확장성을 위해 모든 Oauth제공자에 대해 OauthSignUpToken 내부에는 Oauth의 식별자를 넣도록 해야겠군요!

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/JwtProvider.java:
```diff
@@ -49,6 +49,13 @@ public String createRefreshToken() {
+        Date now = new Date();
+        Date oauthMillis = new Date(now.getTime() + accessExpirationMillis);
+        return buildToken(providerMemberId, now, oauthMillis);
+
```
p3: 개행 없애도 될 것 같아요~

kimdevv commented on backend/fittoring/src/main/resources/db/migration/V202509301652__add_soft_delete_column_to_member_oauth.sql:
```diff
@@ -0,0 +1,5 @@
+ADD COLUMN is_deleted BOOLEAN NOT NULL DEFAULT FALSE;
+
+ALTER TABLE member_oauth
+ADD COLUMN deleted_at DATETIME NULL;
```
```SQL
ALTER TABLE member_oauth
ADD COLUMN is_deleted BOOLEAN NOT NULL DEFAULT FALSE,
ADD COLUMN deleted_at DATETIME NULL;
```

p3: 이렇게 배치 처리를 하면 두 번의 SQL이 아니라 한 번의 SQL으로 끝낼 수 있어서 실행 비용만큼 단축시킬 수 있답니다

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/AuthService.java:
```diff
@@ -96,4 +111,46 @@ private Member createMember(SignUpRequest request) {
+        return memberOauth;
+    }
+
+    public AuthTokenResponse loginOauthMember(MemberOauth memberOauth) {
+        Member member = memberOauth.getMember();
+        return getAuthorizedTokenResponse(member);
+    }
```
p3: 이 부분은 OAuth로 회원가입 할 때만 쓰일 거 같은데 `registerOauthMember()` 메서드와 합치는 건 어떨까요? 위에 `kakaoLogin()`처럼 저마다의 OAuth 로그인 로직이 분리될 것 같습니다.

kimdevv reviewed (commented):
고생하셨습니다~~ 커멘트 남겨드렸으니 확인 부탁드려요!

jye326 committed:
feat: 카카오 OAuth 설정값 테스트 프로필에 추가

- `application-test.yml`에 `kakao.redirect-uri` 및 `kakao.client-id` 설정 추가
- 환경 변수 활용하여 동적 설정 가능하도록 적용

jye326 committed:
ci: 카카오 OAuth 환경 변수 추가

- `KAKAO_REDIRECT_URI` 및 `KAKAO_CLIENT_ID` 환경 변수 추가

ppparkta commented on .github/workflows/backend-ci.yml:
```diff
@@ -59,7 +59,8 @@ jobs:
           echo "ACCESS_TOKEN_EXPIRATION_MILLIS=${{ secrets.ACCESS_TOKEN_EXPIRATION_MILLIS }}" >> .env
           echo "REFRESH_TOKEN_EXPIRATION_MILLIS=${{ secrets.REFRESH_TOKEN_EXPIRATION_MILLIS }}" >> .env
-
+          echo "KAKAO_REDIRECT_URI=https://api.fittoring.com/kakao/callback" >> .env
```
p3: 깃허브 환경변수, 노션 환경변수 잘 업데이트 해주셨네요 👍

ppparkta commented on backend/fittoring/build.gradle:
```diff
@@ -81,6 +81,9 @@ dependencies {
     implementation 'com.github.gotson.nightmonkeys:imageio-heif:1.1.0'
+
+    // String util
+    implementation 'org.apache.commons:commons-lang3:3.14.0'
```
p3: 이건 무슨 기능을 위해 추가된 라이브러리인가요?

jye326 replied:
Oauth 회원가입 멤버에게 할당할 랜덤 아이디, 비밀번호 생성을 위해 추가한 유틸 라이브러리입니다!

ppparkta commented on backend/fittoring/src/main/resources/db/migration/V202509301652__add_soft_delete_column_to_member_oauth.sql:
```diff
@@ -0,0 +1,5 @@
+ADD COLUMN is_deleted BOOLEAN NOT NULL DEFAULT FALSE;
+
+ALTER TABLE member_oauth
+ADD COLUMN deleted_at DATETIME NULL;
```
p3: 아직 반영되기 전이니 말씀드려보면 `V202509300951__create_member_oauth.sql` 파일에 이 파일의 내용이 합쳐져도 되지 않을까요? 제약조건에 대한 수정이 아니라 컬럼 추가라서 두 파일을 분리해서 관리했을 때 이점이 없을 것 같아요

jye326 replied:
소프트 삭제 칼럼 넣는 걸 깜빡했다가 나중에 수정하자니 '절대 플라이 웨이 파일을 수정해선 안돼' 라는 지침이 떠올라 새로 팠습니다😅 말씀하신 대로 반영되기 전이니 합쳐서 올립니다~

ppparkta commented on backend/fittoring/src/test/java/fittoring/mentoring/business/service/AuthServiceTest.java:
```diff
@@ -20,23 +21,34 @@
     private AuthService authService;
 
+    @Autowired
+    OauthClientService oauthClientService;
```
p3: `@MockitoBean` 사용하면 좋지 않을까요? 테스트 코드 상에서 oauthClientService를 사용하는 부분이 없어보이네요!

jye326 replied:
좋네요! 반영했습니다!

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/presentation/dto/OauthSignUpRequest.java:
```diff
@@ -0,0 +1,16 @@
+        String gender,
+        @PhoneNumber
+        @NotBlank
+        String phone) {
```
p2: 개행부탁드려요

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/presentation/api/AuthController.java:
```diff
@@ -65,27 +61,69 @@ public ResponseEntity<Void> reissue(
+        return ResponseEntity.status(HttpStatus.CREATED)
+                .build();
+    }
+
```
p2: 개행 제거해주세요

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/presentation/api/AuthController.java:
```diff
@@ -65,27 +61,69 @@ public ResponseEntity<Void> reissue(
+
+        if (error != null) {
+            return ResponseEntity.status(HttpStatus.BAD_REQUEST)
+                    .body("카카오 로그인 실패 : " + error + " : " + errorDescription);
```
p2: 직접 예외에 대한 ResponseEntity를 처리해주신 이유가 있을까요? 예외를 발생시켜서 기존에 사용하던 ControllerAdvide를 활용하면 예외 양식이 통일돼서 더 보기좋게 반환될 것 같아요

jye326 replied:
반영했습니다!

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/presentation/api/AuthController.java:
```diff
@@ -65,27 +61,69 @@ public ResponseEntity<Void> reissue(
+        }
+
+        if (code == null) {
+            return ResponseEntity.status(HttpStatus.BAD_REQUEST).build();
```
p2: 여기도요!

jye326 replied:
반영했습니다!

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/presentation/api/AuthController.java:
```diff
@@ -65,27 +61,69 @@ public ResponseEntity<Void> reissue(
+        if (code == null) {
+            return ResponseEntity.status(HttpStatus.BAD_REQUEST).build();
+        }
+
+        AuthTokenResponse authTokenResponse = authService.kakaoLogin(code);
+
+        if (authTokenResponse.isLoginSuccess()){
+            CookieWriter.write(httpResponse, authTokenResponse);
+            return ResponseEntity.status(HttpStatus.OK).build();
+        }
+
+        ResponseCookie oauthCookie = CookieProvider.createCookie("oauthSignUpToken", authTokenResponse.oauthSignUpToken());
+        httpResponse.addHeader(HttpHeaders.SET_COOKIE, oauthCookie.toString());
+        return ResponseEntity.status(HttpStatus.MOVED_PERMANENTLY).build();
```
p2: 이 부분은 로그인에 성공하면(=계정이 있으면) 로그인하고, 없으면 회원가입으로 이어지는 로직인가요? 컨트롤러에서 비즈니스 로직을 처리하고 있는 것으로 보여서, Usecase...? Service 레이어로 분리해주시면 좋을 것 같습니다!

jye326 replied:
로그인 성공 시 성공 응답 반환, 계정 정보 없을 시 회원가입 페이지로 리다이렉트 응답을 반환합니다! 회원 가입 로직은 별도의 메서드(oauthSignUp)에서 처리합니다. 

비즈니스 로직이라기 보다는 로그인 성공, 실패 여부에 따라 http응답 쿠키와 응답 상태 코드를 분기해야하는 상황이라 컨트롤러에서 처리하였습니다!

ppparkta reviewed (commented):
보안 이슈를 충분히 인지하고 작업해주신 것 같네요 👍 
코멘트 몇 개 남겼습니다. 고생 많으셨어요!

jye326 committed:
feat: OAuth 회원가입 API oauthSignUpToken 쿠키 처리로 변경

- `OauthSignUpRequest` DTO에서 `oauthSignUpToken` 필드 제거
- `AuthController`에서 `@CookieValue`로 `oauthSignUpToken` 처리
- `registerOauthMember` 메서드에 `oauthSignUpToken` 파라미터 추가
- OAuth 회원가입 요청 시 쿠키를 통한 토큰 전달 방식 적용

jye326 committed:
feat: `member_oauth` 테이블 컬럼 수정

- `is_deleted` 컬럼 및 기본값 `FALSE` 추가
- `deleted_at` 컬럼 추가 및 NULL 허용
- 기존 테이블 생성 SQL 및 마이그레이션 SQL 파일 수정

jye326 committed:
feat: 카카오 OAuth 로그인 및 회원가입 기능 개선

- `AuthProvider` enum 추가로 OAuth 제공자 상수 관리
- `MemberOauth` 엔티티 `provider` 필드를 `String`에서 `AuthProvider` 타입으로 변경
- `kakaoLogin` 메서드에서 액세스 토큰 변수명 명확화 (`accessToken` → `kakaoAccessToken`)
- 예외 처리 개선: `OauthLoginException` 클래스를 생성하여 인증 코드 및 카카오 로그인 에러 처리
- 글로벌 예외 핸들러에 `OauthLoginException` 처리 로직 추가
- DTO, 컨트롤러 및 리포지토리에서 OAuth 관련 코드 리팩터링
- 테스트 코드에 `@MockitoBean`을 사용해 OAuth 클라이언트 설정

jye326 committed:
Merge branch 'develop' into feature/801-ouath-server

# Conflicts:
#	backend/fittoring/src/main/java/fittoring/exception/GlobalExceptionHandler.java
#	backend/fittoring/src/main/resources/application.yml
#	backend/fittoring/src/test/resources/application-test.yml

jye326 committed:
chore: docker-compose 포트 번호 수정

- MySQL 포트 매핑 변경 (13306 → 3306)
- 추가 서비스 포트 매핑 변경 (13307 → 3307)

kimdevv reviewed (approved):
빠르게 반영해 주셨네요. 이만 승인합니다~

ppparkta reviewed (approved):
리뷰 잘 반영되었네요 😄  고생하셨어요. 이만 승인합니다.
